File: code/poisson/poisson_mpi.f
c     a parallel code for Poisson equation 

c     contact seyong.kim81@gmail.com for comments and questions 

      program poisson

      parameter(MAX=1000,SUBMAX=500,IMAX=10000)

      include 'mpif.h'

      integer istatus(MPI_STATUS_SIZE)

      real u(0:(MAX+1),0:(SUBMAX+1)), unew(0:(MAX+1),0:(SUBMAX+1)), 
     #     rho(0:(MAX+1),0:(SUBMAX+1))
      real recvbuf(MAX)

      double precision unorm, usum, resid

      call MPI_INIT(ierr)

      call MPI_COMM_RANK(MPI_COMM_WORLD,node,ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD,numtask,ierr)

      if (node .ne. (numtask-1)) nextup = node + 1

      if (node .ne. 0) nextdn = node - 1

      h = 0.01
      resid = 0.01

c      write(6,*) 'step size = '
c      read(5,*) h
c      write(6,*) 'residual = '
c      read(5,*) resid

      hsq = h*h

      isub = MAX/numtask

      do j = 0, isub+1
         do i = 0, MAX+1
            u(i,j) = 0.0
            rho(i,j) = 0.0
         enddo
      enddo

      do j = 0, isub+1
         u(0,j) = 10.0
      enddo

      itag1 = 22

      itag2 = 33
         
      do iter = 1, IMAX

         do j = 1, isub
            do i = 1, MAX
               unew(i,j) = 0.25*(u(i-1,j)+u(i+1,j)+u(i,j-1)+u(i,j+1)
     #              - hsq*rho(i,j))
            enddo
         enddo

         unorm = 0.0
         do j = 1, isub
            do i = 1, MAX
               unorm = unorm + (unew(i,j)-u(i,j))
     #              *(unew(i,j)-u(i,j))
            enddo
         enddo

         call MPI_ALLREDUCE(unorm,usum,1,MPI_DOUBLE_PRECISION,MPI_SUM,
     #        MPI_COMM_WORLD,ierr)

c         write(6,*) 'node = ', node, '  unorm = ', unorm, ' usum =',
c     #        usum

         if (node .eq. 0) write(6,*) 'loop count = ', iter, ' usum =',
     #        usum

         if(sqrt(usum) .le. sqrt(resid)) then 
            goto 1000
         endif

         do j = 1, isub
            do i = 1, MAX
               u(i,j) = unew(i,j)
            enddo
         enddo

         if (mod(node,2) .eq. 1) then

            call MPI_SEND(unew(1,1),MAX,MPI_REAL,nextdn,itag1,
     #           MPI_COMM_WORLD,ierr1)
            call MPI_RECV(u(1,0),MAX,MPI_REAL,nextdn,itag2,
     #           MPI_COMM_WORLD,istatus,ierr2)

            if (node .lt. (numtask-1)) then
               call MPI_SEND(unew(1,isub),MAX,MPI_REAL,nextup,itag1,
     #              MPI_COMM_WORLD,ierr1)
               call MPI_RECV(u(1,isub+1),MAX,MPI_REAL,nextup,itag2,
     #              MPI_COMM_WORLD,istatus,ierr2)
            endif
            
         else

            if (node .gt. 0) then
               call MPI_RECV(u(1,0),MAX,MPI_REAL,nextdn,itag1,
     #              MPI_COMM_WORLD,istatus,ierr1)
               call MPI_SEND(unew(1,1),MAX,MPI_REAL,nextdn,itag2,
     #           MPI_COMM_WORLD,ierr2)
            endif

            if (node .lt. (numtask-1)) then
               call MPI_RECV(u(1,isub+1),MAX,MPI_REAL,nextup,itag1,
     #              MPI_COMM_WORLD,istatus,ierr1)
               call MPI_SEND(unew(1,isub),MAX,MPI_REAL,nextup,itag2,
     #              MPI_COMM_WORLD,ierr2)
            endif

         endif

      enddo

 1000 continue

c      do i = 0, isub+1
c         write(20+node,*) u(i,MAX/2)
c      enddo

      do i = 0, isub+1
         write(40+node,*) u(10,i)
      enddo

      write(10+node,*) ((u(i,j),i=0,isub+1),j=0,isub+1)

      call MPI_FINALIZE(ierr)

      stop
      end

File: code/poisson/poisson_mpi.c
/* a parallel code for Poisson equation */

/* contact seyong.kim81@gmail.com for comments and questions */


#include <stdio.h>
#include <math.h>
#include <mpi.h>

#define MAX 1000
#define SUBMAX 500
#define IMAX 1000

main(int argc, char** argv) {

  int i, j, flag, node, numtask, nextdn, nextup, isub, itag1, itag2, loop;
  float u[MAX+2][SUBMAX+2], unew[MAX+2][SUBMAX+2], rho[MAX+2][SUBMAX+2];
  float sendbuf[MAX],recvbuf[MAX], h, hsq;
  double unorm, usum, resid;

  MPI_Status istatus;

  FILE *fopen(), *fp12, *fp14, *fp10;

  MPI_Init(&argc,&argv);
  MPI_Comm_rank(MPI_COMM_WORLD,&node);
  MPI_Comm_size(MPI_COMM_WORLD,&numtask);

  if(node != (numtask-1)) nextup = node+1;
  if(node != 0) nextdn = node-1;

  h = 0.01;
  resid = 0.01;
  loop = 0;

  /*  printf("step size = \n"); scanf("%f",&h);
      printf("step size = %f\n",h);*/

  /*  printf("residual = \n"); scanf("%lf",&resid);
      printf("resid = %.7f\n",resid);*/
  
  fp12 = fopen("c.12","w");
  fp14 = fopen("c.14","w");
  fp10 = fopen("c.10","w");

  hsq = h*h;

  isub = MAX/numtask;

  for(j = 0;j <= isub+1;j++)
    for(i = 0;i <= MAX+1;i++) {
      u[i][j] = 0.0;
      rho[i][j] = 0.0;
    }
  for(j = 0;j <= isub+1;j++) u[0][j] = 10.0;

  flag = 1;

  itag1 = 22;

  itag2 = 33;

  while(flag) {
    loop++;
    for(j = 1;j <= isub;j++)
      for(i = 1;i <= MAX;i++)
	unew[i][j] = 0.25*(u[i-1][j]+u[i+1][j]+u[i][j-1]+u[i][j+1]
			   - hsq*rho[i][j]);

    unorm = 0.0;
    for(j = 1;j <= isub;j++)
      for(i = 1;i <= MAX;i++)
	unorm += (unew[i][j]-u[i][j])*(unew[i][j]-u[i][j]);

    MPI_Allreduce(&unorm,&usum,1,MPI_DOUBLE,MPI_SUM,MPI_COMM_WORLD);

    if (node == 0) printf("loop = %d, unorm = %.8e\n",loop,usum);

    if (sqrt(usum) <= sqrt(resid)) flag = 0;

    for(j = 1;j <= isub;j++)
      for(i = 1;i <= MAX;i++) u[i][j] = unew[i][j];
    
    if ((node%2) == 1) {
      for(i=0;i < MAX;i++) sendbuf[i] = unew[i+1][1];
      MPI_Send(sendbuf,MAX,MPI_FLOAT,nextdn,itag1,MPI_COMM_WORLD);
      MPI_Recv(recvbuf,MAX,MPI_FLOAT,nextdn,itag2,MPI_COMM_WORLD,&istatus);
      for(i=0;i < MAX;i++) u[i+1][0] = recvbuf[i];
      if ( node != (numtask-1)) {
	for(i=0;i < MAX;i++) sendbuf[i] = unew[i+1][isub];
	MPI_Send(sendbuf,MAX,MPI_FLOAT,nextup,itag1,MPI_COMM_WORLD);
	MPI_Recv(recvbuf,MAX,MPI_FLOAT,nextup,itag2,MPI_COMM_WORLD,&istatus);
	for(i=0;i < MAX;i++) u[i+1][isub+1] = recvbuf[i];
      }
    }
    else {
      if (node != 0) {
	MPI_Recv(recvbuf,MAX,MPI_FLOAT,nextdn,itag1,MPI_COMM_WORLD,&istatus);
	for(i=0;i < MAX;i++) u[i+1][0] = recvbuf[i];
	for(i=0;i < MAX;i++) sendbuf[i] = unew[i+1][1];
	MPI_Send(sendbuf,MAX,MPI_FLOAT,nextdn,itag2,MPI_COMM_WORLD);
      }

      if (node != (numtask-1)) {
	MPI_Recv(recvbuf,MAX,MPI_FLOAT,nextup,itag1,MPI_COMM_WORLD,&istatus);
	for(i=0;i < MAX;i++) u[i+1][isub+1] = recvbuf[i];
	for(i=0;i < MAX;i++) sendbuf[i] = unew[i+1][isub];
	MPI_Send(sendbuf,MAX,MPI_FLOAT,nextup,itag2,MPI_COMM_WORLD);
      }
    }
  } 

  for(i = 0;i <= isub+1;i++) fprintf(fp12,"%f\n",u[i][MAX/2+1]);
  for(i = 0;i <= isub+1;i++) fprintf(fp14,"%f\n",u[10+1][i]);

  for(j = 0;j <= isub+1;j++)
    for(i = 0;i <= MAX+1;i++) fprintf(fp10,"%f\n",u[i][j]);
  
  MPI_Finalize();
}
